abcix/adapter/adapter.go

```
@@ -32,59 +35,66 @@ func (app *adaptedApp) OriginalApp() abci.Application {
 	return app.abciApp
 }
 
-func (app *adaptedApp) Info(req abcix.RequestInfo) (resp abcix.ResponseInfo) {
-	abciReq := abci.RequestInfo{}
-	if err := copier.Copy(&abciReq, &req); err != nil {
+var typeRegistry = map[string]reflect.Type{
+	"abci.RequestInfo":               reflect.TypeOf(abci.RequestInfo{}),
+	"abci.RequestSetOption":          reflect.TypeOf(abci.RequestSetOption{}),
+	"abci.RequestQuery":              reflect.TypeOf(abci.RequestQuery{}),
+	"abci.RequestCheckTx":            reflect.TypeOf(abci.RequestCheckTx{}),
+	"abci.RequestInitChain":          reflect.TypeOf(abci.RequestInitChain{}),
+	"abci.RequestListSnapshots":      reflect.TypeOf(abci.RequestListSnapshots{}),
+	"abci.RequestOfferSnapshot":      reflect.TypeOf(abci.RequestOfferSnapshot{}),
+	"abci.RequestLoadSnapshotChunk":  reflect.TypeOf(abci.RequestOfferSnapshot{}),
+	"abci.RequestApplySnapshotChunk": reflect.TypeOf(abci.RequestApplySnapshotChunk{}),
+	"abci.RequestEndBlock":           reflect.TypeOf(abci.RequestEndBlock{}),
+	"abci.RequestBeginBlock":         reflect.TypeOf(abci.RequestBeginBlock{}),
+}
+
+func Apply(f interface{}, args ...interface{}) reflect.Value {
+	fun := reflect.ValueOf(f)
+	in := make([]reflect.Value, len(args))
+	for k, arg := range args {
+		in[k] = reflect.ValueOf(arg).Elem()
+	}
+	r := fun.Call(in)
+	return r[0]
+}
+
+func convert(req interface{}, resp interface{}, abciReq string, f interface{}) error {
+	elem, ok := typeRegistry[abciReq]
+	if !ok {
+		return errors.New("fail to build a new struct")
+	}
+	abcireq := reflect.New(elem).Interface()
+	if err := copier.Copy(abcireq, req); err != nil {
 		// TODO: panic for debugging purposes. better error handling soon!
 		panic(err)
```

should return the error